File: LEETCODE/LEETCODE-CPP/45.cpp
//45. Jump Game II
//https://leetcode.com/problems/jump-game-ii/

class Solution {
public:
    //Method-1 (DP solution) -----TLE-----
    int jump(vector<int>& nums) {
        int n=nums.size();
        vector<int>dp(n,INT_MAX);
        
        dp[0]=0;
        
        for(int i=1;i<n;i++)
            for(int j=0;j<i;j++)
                if(j+nums[j]>=i)
                    if(dp[j]!=INT_MAX)
                        dp[i]=min(dp[i],dp[j]+1);
        
        return dp[n-1]==INT_MAX? -1:dp[n-1];
    }
};

File: LEETCODE/LEETCODE-PYTHON/45.py
# 45. Jump Game II
# https://leetcode.com/problems/jump-game-ii/description/

# Method-1 using greedy
class Solution:
    def jump(self, nums: List[int]) -> int:
        n, ans= len(nums),0
        curr_end, curr_far=0,0

        for i in range(0,n-1):
            if curr_end>=n-1:
                break

            curr_far=max(curr_far,i+nums[i])

            if i==curr_end:
                ans+=1
                curr_end=curr_far

        return ans